src/ansi-c/c_nondet_symbol_factory.cpp

```diff
@@ -0,0 +1,419 @@
+/*******************************************************************\
+
+Module: C Nondet Symbol Factory
+
+Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#include <set>
+#include <sstream>
+
+#include <util/arith_tools.h>
+#include <util/fresh_symbol.h>
+#include <util/std_types.h>
+#include <util/std_code.h>
+#include <util/std_expr.h>
+#include <util/namespace.h>
+#include <util/pointer_offset_size.h>
+#include <util/prefix.h>
+
+#include <linking/zero_initializer.h>
+
+#include <ansi-c/c_types.h>
+#include <ansi-c/string_constant.h>
+
+#include <goto-programs/goto_functions.h>
+
+#include "c_nondet_symbol_factory.h"
+
+/*******************************************************************\
+
+Function: c_new_tmp_symbol
+
+ Inputs:
+  symbol_table - The symbol table to create the symbol in
+  loc - The location to assign to the symbol
+  type - The type of symbol to create
+  prefix - The prefix to use for the symbol's basename
+
+ Outputs: Returns a reference to the new symbol
+
+ Purpose: Create a new temporary static symbol
+
+\*******************************************************************/
+
+static symbolt &c_new_tmp_symbol(
+  symbol_tablet &symbol_table,
+  const source_locationt &loc,
+  const typet &type,
+  const std::string &prefix="tmp")
+{
+  symbolt &tmp_symbol=
+    get_fresh_aux_symbol(type, "", prefix, loc, ID_C, symbol_table);
+  tmp_symbol.is_static_lifetime=true;
+  return tmp_symbol;
+}
+
+/*******************************************************************\
+
+Function: c_get_nondet_bool
+
+ Inputs: Desired type (C_bool or plain bool)
+
+ Outputs: nondet expr of that type
+
+ Purpose:
+
+\*******************************************************************/
+
+static exprt c_get_nondet_bool(const typet &type)
+{
+  // We force this to 0 and 1 and won't consider
+  // other values.
+  return typecast_exprt(side_effect_expr_nondett(bool_typet()), type);
+}
+
+/*******************************************************************\
+
+Function: concatenate_code_blocks
+
+ Inputs:
+  target_block - The code_blockt to add the code to
+  extra_block - The code_blockt containing the code to add
+
+ Outputs:
+
+ Purpose: Add all the codets from extra_block to target_block
+
+\*******************************************************************/
+
+void concatenate_code_blocks(
+  code_blockt &target_block,
+  const code_blockt &extra_block)
+{
+  for(const auto &operand : extra_block.operands())
+  {
+    target_block.add(to_code(operand));
+  }
+}
```

Comment:
As suggested in your comment, this should really go in util/std_code.h, as a member function of `code_blockt`. It would then be worthwhile using `reserve` to do the re-allocation of memory just once.